File: backend/natures.py
"""Geek Cards TCG Natures System - single source of truth for weakness chain."""

NATURES = [
    "Anjo", "Demônio", "Entidade", "Mago", "Cavaleiro", "Caçador",
    "Monstro", "Super", "Agente", "Pirata", "Herói", "Vilão",
    "Shinobi", "Dragão"
]

# Weakness chain: key is weak TO the values in the list
WEAKNESS_MAP = {
    "Anjo": [],
    "Demônio": ["Anjo"],
    "Entidade": ["Demônio"],
    "Mago": ["Entidade"],
    "Cavaleiro": ["Mago", "Dragão"],
    "Caçador": ["Cavaleiro"],
    "Monstro": ["Caçador"],
    "Super": ["Monstro"],
    "Agente": ["Super"],
    "Pirata": ["Agente"],
    "Herói": ["Pirata"],
    "Vilão": ["Herói"],
    "Shinobi": ["Vilão"],
    "Dragão": ["Shinobi"],
}

# Who each nature has advantage against (inverse of weakness map)
ADVANTAGE_MAP = {n: [] for n in NATURES}
for nature, weak_to in WEAKNESS_MAP.items():
    for w in weak_to:
        ADVANTAGE_MAP[w].append(nature)


def compute_effective_weaknesses(card_natures: list[str]) -> list[str]:
    """Given a card's natures, compute final weaknesses after cancellations.
    A weakness is cancelled if another characteristic on the card has that weakness's own weakness.
    Simpler rule: take union of all weaknesses, remove any nature that is also in the card's natures.
    """
    if not card_natures:
        return []
    weaknesses = set()
    for n in card_natures:
        for w in WEAKNESS_MAP.get(n, []):
            weaknesses.add(w)
    # Remove weaknesses that are themselves in the card's own natures (self-cancel)
    weaknesses -= set(card_natures)
    return sorted(weaknesses)


CARD_TYPES = ["Personagem", "Item", "Mestre", "Energia"]
ENERGY_TYPES = ["Superior", "Natural", "Interior", "Universal"]
RARITIES = [1, 2, 3]

File: frontend/src/lib/natures.js
// Natures constants - must match backend
export const NATURES = [
  "Anjo", "Demônio", "Entidade", "Mago", "Cavaleiro", "Caçador",
  "Monstro", "Super", "Agente", "Pirata", "Herói", "Vilão",
  "Shinobi", "Dragão"
];

export const NATURE_COLORS = {
  "Anjo": "#FDE047",
  "Demônio": "#991B1B",
  "Entidade": "#7E22CE",
  "Mago": "#3B82F6",
  "Cavaleiro": "#94A3B8",
  "Caçador": "#4D7C0F",
  "Monstro": "#22C55E",
  "Super": "#F59E0B",
  "Agente": "#0EA5E9",
  "Pirata": "#14B8A6",
  "Herói": "#4338CA",
  "Vilão": "#BE185D",
  "Shinobi": "#1F2937",
  "Dragão": "#DC2626",
};

export const WEAKNESS_MAP = {
  "Anjo": [],
  "Demônio": ["Anjo"],
  "Entidade": ["Demônio"],
  "Mago": ["Entidade"],
  "Cavaleiro": ["Mago", "Dragão"],
  "Caçador": ["Cavaleiro"],
  "Monstro": ["Caçador"],
  "Super": ["Monstro"],
  "Agente": ["Super"],
  "Pirata": ["Agente"],
  "Herói": ["Pirata"],
  "Vilão": ["Herói"],
  "Shinobi": ["Vilão"],
  "Dragão": ["Shinobi"],
};

export const CARD_TYPES = ["Personagem", "Item", "Mestre", "Energia"];
export const ENERGY_TYPES = ["Superior", "Natural", "Interior", "Universal"];

export const RARITY_COLORS = {
  0: "#94A3B8",
  1: "#abffa8",
  2: "#4e91fd",
  3: "#c83cff",
  4: "#ff2f2f",
  alpha: "#ffd900",
};

export function computeEffectiveWeaknesses(natures) {
  if (!natures || natures.length === 0) return [];
  const weak = new Set();
  for (const n of natures) {
    (WEAKNESS_MAP[n] || []).forEach(w => weak.add(w));
  }
  natures.forEach(n => weak.delete(n));
  return [...weak].sort();
}
